kamu-core/src/infra/sync_service_impl.rs

```diff
@@ -241,20 +243,46 @@ impl SyncServiceImpl {
             AppendValidation::Full
         };
 
-        info!("Starting sync using Simple Transfer Protocol");
-        match SimpleTransferProtocol
-            .sync(
+        let dst_dataset = dst_dataset_builder.as_dataset();
+
+        let sync_result = if src_dataset.supports_smart_protocol() {
+            self.sync_smart_pull_transfer_protocol(
                 src_dataset.as_ref(),
                 src,
-                dst_dataset_builder.as_dataset(),
+                dst_dataset,
                 dst,
                 validation,
                 opts.trust_source.unwrap_or(src_is_local),
                 opts.force,
                 listener,
-            )
-            .await
-        {
+            ).await
+
+        } else if dst_dataset.supports_smart_protocol() {
```

Comment:
Perhaps `sync_generic` should not exist any more - it's not really "generic" if it just branches based on protocol and the operation. 

In the method where we analyze the dataset URL we probably should pick the specific protocol implementation to use (function pointer / strategy object).

Reply:
I'm analyzing URLs to select an algorithm, but I've left `sync_generic` untouched - it's a default case, when ODF refs are not detected